### ISSUE TITLE
deadlock: investigate deadlock problem

### ISSUE THREAD

**Issue by kristianlm** [Adellica/chicken-gochan#1]
Running this:

``` scheme
;;; stress-test gochan. This is causing deadlocks after about 25000
;;; messages (consistantly). We need to figure out what's going wrong here!
(use gochan)

(define chan (gochan))

(thread-start!
 (lambda () ;; simple echo
   (gochan-for-each
    chan (lambda (msg)
           (gochan-send (car msg)
                        (cdr msg))))))

(let loop ((n 100000))
  (if (> n 0)
      (let ((reply (gochan)))
        (gochan-send chan (cons reply n))
        (assert (eq? n (gochan-receive reply)))
        (loop (sub1 n)))))
```

We get a deadlock error after about 25000 loops when interpreted, 400 loops when compiled with `-O5` and never when compiled with `-O1`. I believe there's a mutex-lock missing somewhere.

**Comment by kristianlm** [comment]
I think the semaphore implementation is broken. [Here](https://github.com/Adellica/chicken-gochan/blob/master/gochan.scm#L44), for example, the semaphore could be closed between `semaphore-open?` and `mutex-unlock!`

The semaphores will need some kind of mutexing too.

**Comment by kristianlm** [comment]
This was fixed in 81a7193e99ce75ba56221390a3fc596f4ef40310
